File: javascript/13_sorts/bucketSort.js
// 思路：
// 将数组中的数据，按桶进行划分，将相邻的数据划分在同一个桶中
// 每个桶用插入排序算法（或者快速排序）进行排序
// 最后整合每个桶中的数据

function bucketSort(array, bucketSize = 5) {
    if (array.length < 2) {
        return array
    }
    const buckets = createBuckets(array, bucketSize)
    return sortBuckets(buckets)
}

function createBuckets(array, bucketSize) {
    let minValue = array[0]
    let maxValue = array[0]
    // 遍历数组，找到数组最小值与数组最大值
    for (let i = 1; i < array.length; i++) {
        if (array[i] < minValue) {
            minValue = array[i]
        } else if (array[i] > maxValue) {
            maxValue = array[i]
        }
    }
    // 根据最小值、最大值、桶的大小，计算得到桶的个数
    const bucketCount = Math.floor((maxValue - minValue) / bucketSize) + 1
    // 建立一个二维数组，将桶放入buckets中
    const buckets = []
    for (let i = 0; i < bucketCount; i++) {
        buckets[i] = []
    }
    // 计算每一个值应该放在哪一个桶中
    for (let i = 0; i < array.length; i++) {
        const bucketIndex = Math.floor((array[i] - minValue) / bucketSize)
        buckets[bucketIndex].push(array[i])
    }
    return buckets
}

function sortBuckets(buckets) {
    const sortedArray = []
    for (let i = 0; i < buckets.length; i++) {
        if (buckets[i] != null) {
            insertionSort(buckets[i])
            sortedArray.push(...buckets[i])
        }
    }
    return sortedArray
}

// 插入排序
function insertionSort(array) {
    const { length } = array
    if (length <= 1) return

    for (let i = 1; i < length; i++) {
        let value = array[i]
        let j = i - 1

        while (j >= 0) {
            if (array[j] > value) {
                array[j + 1] = array[j] // 移动
                j--
            } else {
                break
            }
        }
        array[j + 1] = value // 插入数据
    }
}

File: kotlin/13_sorts/BucketSort.kt
/**
 * @Description:桶排序算法
 * @Author: Hoda
 * @Date: Create in 2019-06-01
 * @Modified By:
 * @Modified Date:
 */
object BucketSort {
    /**
     * 桶排序
     *
     * @param arr 数组
     * @param bucketSize 桶容量
     */
    fun bucketSort(arr: IntArray, bucketSize: Int) {
        if (arr.size < 2) {
            return
        }
        // 数组最小值
        var minValue = arr[0]
        // 数组最大值
        var maxValue = arr[1]
        for (i in arr.indices) {
            if (arr[i] < minValue) {
                minValue = arr[i]
            } else if (arr[i] > maxValue) {
                maxValue = arr[i]
            }
        }
        // 桶数量
        val bucketCount = (maxValue - minValue) / bucketSize + 1
        val buckets =
            Array(bucketCount) { IntArray(bucketSize) }
        val indexArr = IntArray(bucketCount)
        // 将数组中值分配到各个桶里
        for (i in arr.indices) {
            val bucketIndex = (arr[i] - minValue) / bucketSize
            if (indexArr[bucketIndex] == buckets[bucketIndex].length) {
                ensureCapacity(buckets, bucketIndex)
            }
            buckets[bucketIndex][indexArr[bucketIndex]++] = arr[i]
        }
        // 对每个桶进行排序，这里使用了快速排序
        var k = 0
        for (i in buckets.indices) {
            if (indexArr[i] == 0) {
                continue
            }
            quickSortC(buckets[i], 0, indexArr[i] - 1)
            for (j in 0 until indexArr[i]) {
                arr[k++] = buckets[i][j]
            }
        }
    }

    /**
     * 数组扩容
     *
     * @param buckets
     * @param bucketIndex
     */
    private fun ensureCapacity(
        buckets: Array<IntArray>,
        bucketIndex: Int
    ) {
        val tempArr = buckets[bucketIndex]
        val newArr = IntArray(tempArr.size * 2)
        for (j in tempArr.indices) {
            newArr[j] = tempArr[j]
        }
        buckets[bucketIndex] = newArr
    }

    /**
     * 快速排序递归函数
     *
     * @param arr
     * @param p
     * @param r
     */
    private fun quickSortC(arr: IntArray, p: Int, r: Int) {
        if (p >= r) {
            return
        }
        val q = partition(arr, p, r)
        quickSortC(arr, p, q - 1)
        quickSortC(arr, q + 1, r)
    }

    /**
     * 分区函数
     *
     * @param arr
     * @param p
     * @param r
     * @return 分区点位置
     */
    private fun partition(arr: IntArray, p: Int, r: Int): Int {
        val pivot = arr[r]
        var i = p
        for (j in p until r) {
            if (arr[j] <= pivot) {
                swap(arr, i, j)
                i++
            }
        }
        swap(arr, i, r)
        return i
    }

    /**
     * 交换
     *
     * @param arr
     * @param i
     * @param j
     */
    private fun swap(arr: IntArray, i: Int, j: Int) {
        if (i == j) {
            return
        }
        val tmp = arr[i]
        arr[i] = arr[j]
        arr[j] = tmp
    }
}